### PR TITLE
Use PluginCleanupPolicy from flyteorg/flyteplugins#203

### PR DESCRIPTION
**Pull request by clairemcginty**
# TL;DR
Per discussion in `#feature-discussions` Slack channel, I'm opening this as a draft PR as to center discussions related to flyteorg/flyte#1345.

This PR applies the `PluginCleanupPolicy` interface from flyteorg/flyteplugins#203, if applicable, instead of defaulting to just deleting the resource when the task is aborted.

this will need to be updated once the flyteplugins changed is merged & its version is bumped in `go.mod`.

cc @regadas , @EngHabu , @kumare3 :)

## Type
 - [ ] Bug Fix
 - [ ] Feature
 - [x] Plugin

## Are all requirements met?

 - [ ] Code completed
 - [ ] Smoke tested
 - [x] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue
_Remove the '*fixes*' keyword if there will be multiple PRs to fix the linked issue_

fixes https://github.com/flyteorg/flyte/issues/<number>

## Follow-up issue
_NA_
OR
_https://github.com/flyteorg/flyte/issues/<number>_

### CODE REVIEW

File: pkg/controller/nodes/task/k8s/plugin_manager.go [hunk 1]
```diff
@@ -335,7 +335,31 @@ func (e PluginManager) Abort(ctx context.Context, tCtx pluginsCore.TaskExecution
...
+	if err != nil {
+	} else if deleteResource {
+		err = e.kubeClient.GetClient().Delete(ctx, o)
+	} else {
+		if behavior.Patch != nil && behavior.Update == nil {
+			err = e.kubeClient.GetClient().Patch(ctx, o, behavior.Patch.Patch, behavior.Patch.Options...)
+		} else if behavior.Update != nil {
+			err = e.kubeClient.GetClient().Update(ctx, o, behavior.Update.Options...)
+		} else {
+			err = errors.Errorf(errors.RuntimeFailure, "AbortBehavior for resource %v must specify either a Patch and an Update operation if Delete is set to false. Only one can be supplied.", o.GetName())
```
Comment (clairemcginty) [inline pkg/controller/nodes/task/k8s/plugin_manager.go:355]: Maybe worth adding a custom error type to flyteplugins' `errors` package? `InvalidPluginConfiguration` or something like that? cc @EngHabu

**Comment by kumare3** [comment]
looks good to me, I will let @EngHabu +1.
Can you sign the dco?

File: pkg/controller/nodes/task/k8s/plugin_manager.go [hunk 2]
```diff
@@ -335,7 +335,31 @@ func (e PluginManager) Abort(ctx context.Context, tCtx pluginsCore.TaskExecution
...
+		deleteResource = err == nil && behavior.DeleteResource
+	}
+
+	if err != nil {
+	} else if deleteResource {
+		err = e.kubeClient.GetClient().Delete(ctx, o)
+	} else {
+		if behavior.Patch != nil && behavior.Update == nil {
+			err = e.kubeClient.GetClient().Patch(ctx, o, behavior.Patch.Patch, behavior.Patch.Options...)
+		} else if behavior.Update != nil {
```
Comment (EngHabu) [inline pkg/controller/nodes/task/k8s/plugin_manager.go:352]:
```suggestion
		} else if behavior.Patch == nil && behavior.Update != nil {
```

File: pkg/controller/nodes/task/k8s/plugin_manager.go [hunk 3]
```diff
@@ -335,7 +335,31 @@ func (e PluginManager) Abort(ctx context.Context, tCtx pluginsCore.TaskExecution
...
 
-	err = e.kubeClient.GetClient().Delete(ctx, o)
+	deleteResource := true
+	abortOverride, hasAbortOverride := e.plugin.(k8s.PluginAbortOverride)
+	var behavior k8s.AbortBehavior
+	if hasAbortOverride {
+		behavior, err = abortOverride.OnAbort(ctx, tCtx, o)
+		deleteResource = err == nil && behavior.DeleteResource
+	}
+
```
Comment (EngHabu) [inline pkg/controller/nodes/task/k8s/plugin_manager.go:345]: We should also use `behavior.Resource` if set instead of `o`

File: pkg/controller/nodes/task/k8s/plugin_manager_test.go [hunk 4]
```diff
@@ -441,6 +514,69 @@ func TestPluginManager_Abort(t *testing.T) {
...
+			})
+
+		assert.NotNil(t, res)
+		assert.NoError(t, err)
+
+		err = pluginManager.Abort(ctx, tctx)
+		assert.Equal(t, expectedErr, err)
+	})
+
+	t.Run("Abort Plugin has Delete PluginAbortOverride", func(t *testing.T) {
```
Comment (clairemcginty) [inline pkg/controller/nodes/task/k8s/plugin_manager_test.go:563]: I didn't include tests for AbortBehaviors with a non-nil `Resource` because the existing test framework made it a little difficult (can't make assertions on the calls made to the `extendedFakeClient` as it's not a mock)... I'll brainstorm a bit if there's an easier way to test that.

**Comment by EngHabu** [comment]
You will need to run `git pull origin master` and resolve conflicts then `git add -A`, `git merge --continue` and `git push`...

**Review by EngHabu — APPROVED** [review]

**Comment by EngHabu** [comment]
🙇🏽
